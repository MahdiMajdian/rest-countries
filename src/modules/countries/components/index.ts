export { default as SearchInput } from './SearchInput';
export { default as CountryCard } from './CountryCard';
