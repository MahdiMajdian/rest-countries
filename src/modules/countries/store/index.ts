export type { HTTPService, Country } from './types';

export { default as useStore } from './store';
export type { Store } from './store';
