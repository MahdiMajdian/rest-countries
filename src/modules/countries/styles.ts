import styled from '@emotion/styled';
import { Link as OriginalLink } from 'react-router-dom';

export const Option = styled.option``;

export const SelectBox = styled.select`
  width: 100%;
  border: none;
  border-radius: 4px;
  outline: none;
  color: ${({ theme }) => theme.palette.primary};
  font-size: 16px;
  font-family: inherit;
  background-color: ${({ theme }) => theme.background.element};
  box-shadow: ${({ theme }) => theme.shadow.main};
  appearance: none;
  padding-inline: 16px;
`;

export const Icon = styled.div`
  display: flex;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  align-items: center;
  padding-inline-end: 16px;
  pointer-events: none;

  svg {
    width: 24px;
    height: 24px;
    fill: ${({ theme }) => theme.palette.primary};
  }
`;

export const FilterByRegion = styled.div`
  display: flex;
  position: relative;
  width: 55%;
  height: 48px;
  margin-block: 16px;
  margin-inline: 16px;

  @media (min-width: ${({ theme }) => theme.breakpoints.desktop}) {
    width: 15%;
    margin: 0;
  }
`;

export const MenuBar = styled.div`
  display: flex;
  flex-direction: column;

  @media (min-width: ${({ theme }) => theme.breakpoints.desktop}) {
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    margin-block-start: 48px;
  }
`;

export const Item = styled.div`
  font-weight: ${({ theme }) => theme.fontWeight.normal};
  font-size: 12px;
  margin-block-end: 12px;
`;

export const Value = styled.span`
  font-weight: ${({ theme }) => theme.fontWeight.light};
  font-size: 12px;
`;

export const Title = styled.p`
  font-weight: ${({ theme }) => theme.fontWeight.bold};
  margin-block-end: 24px;
`;

export const Information = styled.div`
  width: 100%;
  padding-block-start: 24px;
  padding-block-end: 32px;
  padding-inline: 24px;
`;

export const Flag = styled.img`
  width: 100%;
  aspect-ratio: 1.7;
  object-fit: cover;
`;

export const Card = styled.li`
  width: 264px;
  margin: auto;
  border-radius: 4px;
  overflow: hidden;
  list-style: none;
  background-color: ${({ theme }) => theme.background.element};
  box-shadow: ${({ theme }) => theme.shadow.main};

  @media (min-width: ${({ theme }) => theme.breakpoints.desktop}) {
    height: 100%;
  }
`;

export const Link = styled(OriginalLink)`
  text-decoration: none;
  color: inherit;
`;

export const CountryList = styled.ul`
  display: flex;
  flex-direction: column;
  margin-block-start: 16px;
  padding-inline: 16px;
  gap: 40px;

  @media (min-width: ${({ theme }) => theme.breakpoints.desktop}) {
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
    padding: 0;
    gap: 66px;
    margin-block-start: 48px;
  }
`;

export const StyledCountries = styled.div`
  display: flex;
  flex-direction: column;
  max-width: ${({ theme }) => theme.breakpoints.desktop};
  margin: auto;
  color: ${({ theme }) => theme.palette.primary};
  background-color: ${({ theme }) => theme.background.default};

  @media (min-width: ${({ theme }) => theme.breakpoints.desktop}) {
    padding-inline: 84px;
  }
`;
