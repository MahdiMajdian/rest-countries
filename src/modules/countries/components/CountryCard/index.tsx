import { Country } from '@modules/countries/store';

import {
  Card,
  Flag,
  Information,
  Item,
  Link,
  StyledCountryCard,
  Title,
  Value,
} from './styles';

interface Props {
  country: Country;
}

function CountryCardComponent({ country }: Props) {
  return (
    <StyledCountryCard>
      <Link to={`/country/${country.id}`}>
        <Card>
          <Flag src={country.flag.src} alt={country.name} loading={'lazy'} />
          <Information>
            <Title>{country.name}</Title>
            <Item>
              Population: <Value>{country.population}</Value>
            </Item>
            <Item>
              Region: <Value>{country.region}</Value>
            </Item>
            <Item>
              Capital: <Value>{country.capital}</Value>
            </Item>
          </Information>
        </Card>
      </Link>
    </StyledCountryCard>
  );
}

export default CountryCardComponent;
