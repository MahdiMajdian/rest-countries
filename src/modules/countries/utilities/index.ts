import { createContext } from 'react';

import { HTTPService, Store } from '../store';

export const StoreContext = createContext<Store>(null!);

export const ServiceContext = createContext<HTTPService>(null!);

export const DEBOUNCE_DELAY = 200;

export const REGIONS = [
  { id: 0, name: 'Filter by Region' },
  { id: 1, name: 'Africa' },
  { id: 2, name: 'Americas' },
  { id: 3, name: 'Asia' },
  { id: 4, name: 'Europe' },
  { id: 5, name: 'Oceania' },
];

export { COUNTRY_NAMES } from './countries-name';
export { ThreeLetterCountryCodeToFullName } from './convert-country-name';
