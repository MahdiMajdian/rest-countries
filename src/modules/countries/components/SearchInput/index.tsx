import { useCallback, useEffect } from 'react';

import { Magnifier } from '@assets/images';
import useQueryString from '@hooks/useQueryString';
import { DEBOUNCE_DELAY } from '@modules/countries/utilities';

import { Input, SearchIcon, StyledSearchInput } from './styles';

interface Props {
  onQueryChange: (query: string) => void;
  onExecute: (query: string) => void;
}

function SearchInputComponent({
  onQueryChange,
  onExecute,
}: Props): React.ReactElement {
  const { value: searchQuery, changeValue: setSearchQuery } = useQueryString(
    'search',
    ''
  );

  const handleSearchQueryChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      setSearchQuery(event.target.value.replaceAll(' ', ''));
    },
    []
  );

  useEffect(() => {
    onQueryChange(searchQuery);

    if (searchQuery === '') {
      return;
    }

    const debounceTimerId = window.setTimeout(() => {
      onExecute(searchQuery);
    }, DEBOUNCE_DELAY);

    return () => {
      clearTimeout(debounceTimerId);
    };
  }, [searchQuery]);

  return (
    <StyledSearchInput>
      <SearchIcon>
        <Magnifier />
      </SearchIcon>
      <Input
        aria-label='search-input'
        value={searchQuery}
        onChange={handleSearchQueryChange}
        placeholder='Search for a country...'
      />
    </StyledSearchInput>
  );
}

export default SearchInputComponent;
