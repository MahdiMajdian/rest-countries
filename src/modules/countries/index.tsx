import React, {
  useContext,
  useEffect,
  useCallback,
  useMemo,
  useState,
} from 'react';

import { observer } from 'mobx-react';

import { ChevronDown } from '@assets/images';
import useQueryString from '@hooks/useQueryString';

import { SearchInput } from './components';
import {
  CountryList,
  Flag,
  Information,
  Card,
  StyledCountries,
  Title,
  Item,
  Value,
  FilterByRegion,
  SelectBox,
  Icon,
  Option,
  Link,
  MenuBar,
} from './styles';
import { filterCountryByRegion, REGIONS, StoreContext } from './utilities';
function Countries(): React.ReactElement {
  const store = useContext(StoreContext);
  const [searchQuery, setSearchQuery] = useState('');
  const { value: selectedRegion, changeValue: setSelectedRegion } =
    useQueryString('region', REGIONS[0].id);

  const handleSearchQueryChange = useCallback((searchQuery: string) => {
    setSearchQuery(searchQuery);
    store.clearSearchedCountries();
  }, []);

  const handleSearchForCountries = useCallback((searchQuery: string) => {
    store.searchForCountries(searchQuery);
  }, []);

  const handleFilterByRegionChange = useCallback(
    (event: React.ChangeEvent<HTMLSelectElement>) => {
      setSelectedRegion(event.target.value);
    },
    []
  );

  const filteredCountries = useMemo(() => {
    if (store.countries === undefined) {
      return undefined;
    }
    return filterCountryByRegion(store.countries, selectedRegion);
  }, [store.countries, selectedRegion]);

  const searchedAndFilteredCountries = useMemo(() => {
    if (store.searchedCountries === undefined) {
      return undefined;
    }
    return filterCountryByRegion(store.searchedCountries, selectedRegion);
  }, [store.searchedCountries, selectedRegion]);

  useEffect(() => {
    store.getCountries();
  }, []);

  return (
    <StyledCountries>
      <MenuBar>
        <SearchInput
          onQueryChange={handleSearchQueryChange}
          onExecute={handleSearchForCountries}
        />

        <FilterByRegion>
          <SelectBox
            onChange={handleFilterByRegionChange}
            value={selectedRegion}
          >
            {REGIONS.map((region) => (
              <Option key={region.id} value={region.id}>
                {region.label}
              </Option>
            ))}
          </SelectBox>

          <Icon>
            <ChevronDown />
          </Icon>
        </FilterByRegion>
      </MenuBar>

      <CountryList>
        {searchQuery === ''
          ? filteredCountries === undefined
            ? 'Loading All Countries...'
            : filteredCountries.map((country) => {
                return (
                  <Link key={country.id} to={`/country/${country.id}`}>
                    <Card>
                      <Flag
                        src={country.flag.src}
                        alt={country.name}
                        loading={'lazy'}
                      />
                      <Information>
                        <Title>{country.name}</Title>
                        <Item>
                          Population: <Value>{country.population}</Value>
                        </Item>
                        <Item>
                          Region: <Value>{country.region}</Value>
                        </Item>
                        <Item>
                          Capital: <Value>{country.capital}</Value>
                        </Item>
                      </Information>
                    </Card>
                  </Link>
                );
              })
          : searchedAndFilteredCountries === undefined
          ? `Searching for ${searchQuery}`
          : searchedAndFilteredCountries &&
            searchedAndFilteredCountries.length === 0
          ? 'Nothing was found!'
          : searchedAndFilteredCountries.map((country) => {
              return (
                <Link key={country.id} to={`/country/${country.id}`}>
                  <Card>
                    <Flag src={country.flag.src} alt='' />
                    <Information>
                      <Title>{country.name}</Title>
                      <Item>
                        Population: <Value>{country.population}</Value>
                      </Item>
                      <Item>
                        Region: <Value>{country.region}</Value>
                      </Item>
                      <Item>
                        Capital: <Value>{country.capital}</Value>
                      </Item>
                    </Information>
                  </Card>
                </Link>
              );
            })}
      </CountryList>
    </StyledCountries>
  );
}

export default observer(Countries);
