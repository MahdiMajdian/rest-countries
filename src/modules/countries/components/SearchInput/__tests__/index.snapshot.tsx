import { SearchInput } from '@modules/countries/components';
import { render } from '@utilities';

describe('SearchInput', () => {
  describe('Snapshot', () => {
    it('renders', () => {
      const { container } = render(
        <SearchInput
          onExecute={() => {
            return;
          }}
          onQueryChange={() => {
            return;
          }}
        />
      );

      expect(container).toMatchSnapshot();
    });
  });
});
