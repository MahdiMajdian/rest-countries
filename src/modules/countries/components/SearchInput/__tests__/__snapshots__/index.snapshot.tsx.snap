// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput Snapshot renders 1`] = `
<div>
  <div
    class="css-cdxmyc"
  >
    <div
      class="css-1nm4ymt"
    >
      <svg>
        magnifier.svg
      </svg>
    </div>
    <input
      aria-label="search-input"
      class="css-155pe6x"
      placeholder="Search for a country..."
      value=""
    />
  </div>
</div>
`;
